cmd: dockerfile resolve

Dentrax opened the pull request:
Fixes #648
Fixes #707

Signed-off-by: Furkan <furkan.turkal@trendyol.com>
Co-authored-by: Batuhan <batuhan.apaydin@trendyol.com>

cc @developer-guy

dlorenc commented on cmd/cosign/cli/dockerfile/resolve.go:
```diff
@@ -0,0 +1,104 @@
+		return fmt.Errorf("failed extracting images from Dockerfile: %w", err)
+	}
+
+	if c.Output != "" {
```
Should we handle these errors?

developer-guy replied:
forget about the errors, they can be fixed 🤩 thank you so much for your support, you are our hero, I still don't understand how do you manage all of your work, amazing effort 🤝👏!!!

Dentrax replied:
Done! Now, we also support `COPY --from=IMAGE` cases. 🎉 

cc: @dekkagaijin

dlorenc reviewed (commented):
One nit.

dekkagaijin commented on cmd/cosign/cli/dockerfile/resolve.go:
```diff
@@ -0,0 +1,108 @@
+				tmp.WriteString(strings.ReplaceAll(line, image, d.String()))
+			}
+		} else if strings.HasPrefix(strings.ToUpper(line), "COPY") {
+
```
pls implement

Dentrax replied:
done! forgot this 🙈 and updated the e2e tests:

```
FROM --platform=linux/amd64 gcr.io/distroless/base@sha256:4f25af540d54d0f43cd6bc1114b7709f35338ae97d29db2f9a06012e3e82aba8 AS base
COPY --from=gcr.io/distroless/base@sha256:4f25af540d54d0f43cd6bc1114b7709f35338ae97d29db2f9a06012e3e82aba8
```

developer-guy commented:
Kindly ping folks. It seems everything is ok right now, which means it is ready to merge. 🙋🏻‍♂️

dlorenc commented:
Will take a look!

imjasonh commented on cmd/cosign/cli/dockerfile/resolve.go:
```diff
@@ -0,0 +1,106 @@
+			default:
+				ref, err := name.ParseReference(image)
+				if err != nil {
+					// we should not return err here since
```
Should we log this at least? Letting these slip through seems like it might give users a false sense of security.

Or, can we add a --build-arg flag that lets us resolve these too? (Probably a TODO for a future PR)

mattmoor replied:
I’d bias towards secure defaults.  Here not resolving is a validation hole, and silently allowing it feels wrong to me.  Whereas if we break, users will report the issue, and we can fix anything we think we should support but don’t.

imjasonh commented on cmd/cosign/cli/dockerfile/resolve_test.go:
```diff
@@ -0,0 +1,101 @@
+	}{
+		{
+			"happy alpine",
+			`FROM alpine:3.13`,
```
Add a test case that has extra spaces that we'll trim out?

mattmoor commented on cmd/cosign/cli/dockerfile.go:
```diff
@@ -106,3 +107,42 @@ Shell-like variables in the Dockerfile's FROM lines will be substituted with val
+		Long: `Resolve the digest of the images and rewrites them with fully qualified image reference
+
+This command creates a surface which resolves mutable image tags into immutable image digests:
+	- FROM <tag>, it rewrites the Dockerfile to FROM <verified digest>
```
I am on my phone, but I don’t see digest verification 🤔

mattmoor commented on test/testdata/with_copy.Dockerfile:
```diff
@@ -0,0 +1,18 @@
+# limitations under the License.
+
+FROM --platform=linux/amd64 gcr.io/distroless/base AS base
+COPY --from=gcr.io/distroless/base
```
Isn’t this line incomplete?

Dentrax commented:
Sorry, we dropped the ball here. I think we resolved all reviews.

Dentrax commented:
Kind ping here for review 🤗

Dentrax commented:
Rebased from main branch and there are some failing checks. I ran `go mod tidy` on my local but pushing the changes breaks the compilation, otherwise goreleaser throws _diff_ error. 🤔

Dentrax commented:
Rebased, fixed the e2e test. Failing action throws the following error but couldn't understand:

```
ERROR: unknown cluster "kind"
```

Can someone please review? Thanks!

Dentrax commented:
Rebased.

ChaosInTheCRD commented on cmd/cosign/cli/dockerfile/resolve.go:
```diff
@@ -0,0 +1,108 @@
+					// we can define the image refs smth like
+					// i.e., FROM alpine:$(TAG), FROM $(IMAGE), etc.
+					// TODO: support parameter substitution by passing a `--build-arg` flag
+					fmt.Fprintf(os.Stderr, "WARNING: parameter substitution for images does not supported yet. consider set all environment variables before resolving the Dockerfile. Image: %s.\n", image)
```
```suggestion
					fmt.Fprintf(os.Stderr, "WARNING: parameter substitution for images is not supported yet. consider set all environment variables before resolving the Dockerfile. Image: %s.\n", image)
```

I would also like to consider that this warning is made a little more clear. I don't have context, but maybe this warning could be slightly better written for usability 😄

ChaosInTheCRD replied:
Also, surely this is not the only scenario where this function could fail?

Dentrax replied:
Thanks!

> only scenario where this function could fail

in case if we couldn't parse the image reference

ChaosInTheCRD commented on cmd/cosign/cli/dockerfile/verify.go:
```diff
@@ -82,6 +85,20 @@ func getImagesFromDockerfile(dockerfile io.Reader) ([]string, error) {
 }
 
 func getImageFromLine(line string) string {
+	if strings.HasPrefix(strings.ToUpper(line), "COPY") {
```
If I were to pass in a Dockerfile like:
```
FROM golang:1.18 AS builder
RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o app .

FROM alpine:latest 
WORKDIR /root/
COPY --from=builder /go/src/github.com/alexellis/href-counter/app .
CMD ["./app"]  
```

it would fail. `builder` is not an image reference, but it will be successfully passed as one with `name.ParseReference` Maybe it would be good to add a unit test to cover this scenario 😄

ChaosInTheCRD replied:
I am pretty sure also that the value after the `AS` could be any kind of string as well (e.g. `build1`), so will make it hard to try and interpret at the `--from` level

Dentrax replied:
Hmmm, actually I expect two image references: `golang:1.18` and `alpine:latest`

>  it will be successfully passed

But this is the expected case? I couldn't get why you expect this would fail. 🤔 

I added this case into unit tests anyway.

ChaosInTheCRD replied:
So after checking out your branch and testing out, I see that I misunderstood and the test case I presented is fine (as reflected in the test case you added). However, the following scenario:

```
COPY --from=nginx:latest /etc/nginx/nginx.conf /nginx.conf
```

results in the `nginx:latest` not getting pinned to a digest.

Dentrax replied:
Hey @ChaosInTheCRD, just pushed a new commit now it supports to resolve the following statement:

```
COPY --from=nginx:latest /etc/nginx/nginx.conf /nginx.conf
```

Thanks!

ChaosInTheCRD commented:
This is great work - but I am wondering whether it makes sense to have functionality like this directly within cosign. Sure it's helpful, but the art of ensuring that pinning images to digest may be better sitting in a different tool, not to mention spans way further than just Dockerfiles.

@dlorenc @Dentrax @mattmoor @imjasonh what are your thoughts on this? Also adding @znewman01 as he has been considering the digest pinning scenario in https://github.com/sigstore/cosign/pull/2313

znewman01 commented:
I would normally agree—this sounds a little "kitchen sink-y" to me. But I think if we have `cosign dockerfile verify` in the first place we need to flesh out the rest of the story here.

IMO reason this *wouldn't* go in another tool like Ratchet would be the *separation* in the tooling between "resolve" and "verify." It'd be way to easy to "resolve" with Ratchet and wind up with image digests that are untrusted.

TBH my ideal flow would be one-step, checking signatures as *part* of resolving: `cosign dockerfile resolve --key mykey.pub Dockerfile.unpinned > Dockerfile`. This is nice because there's no intermediate resolved-but-unverified Dockerfile to accidentally build with.

Dentrax commented:
Add a support to resolve `--from=image:tag` cases. Also rebased.

znewman01 commented:
Thanks! Looking again at #648 I see that it proposes that the `resolve` command map to `FROM <verified digest>`.

I think we probably want to make this do all the verification of `verify`, just with different output.
